### DISCUSSION TITLE
Import md/html with assets

### DISCUSSION THREAD

**Discussion by thumDer** [outline/outline#5092] [Ideas]
It would be a cool feature, to allow importing md or html files with all their referenced assets (images first of all). I think the way [nuclino](https://blog.nuclino.com/new-in-nuclino-data-import-improved-editor-and-more) handles it for html's is good enough: imorting a zip file with where the md / html file and assets are boundled together. It should work for multiple pages in a zip also.

**Comment by tommoor** [comment]
You can definitely do this already if you get the format just right. The best way to check this would be to do an Export as Markdown from within Outline and then follow that format for data you want to import.

**Reply by thumDer** [reply]
So you mean this is only possible from Preferences / Import / Markdown only, right? When creating a new page from the context menu, zip files are not allowed. 
![kép](https://user-images.githubusercontent.com/54556708/228247179-71503aec-6b95-4436-8b46-d2e3b71bf0d5.png)
But since the functionality already exists under the hood how much an extra effort would it be to extend the single page import to handle zip files? It could be a low hanging fruit.

**Reply by thumDer** [reply]
I've tried importing from preferences, but it seems generic image files (png, jpg) are not processed as an asset, but as a page, and their binary contents gets extracted.
I've checked the exported content, and there you have all assets as extensionless `imported` files, so the export itself is also invalid, because the img links in the md files are pointing to non-existing files. Should I report an issue on this, or is this the expected behaviour?
If so, then creating such md folder manually is not very effective, or even possible.
After all it seems my original request is still valid more or less.

**Reply by thumDer** [reply]
Just a slight clarification: the above issue is only valid for pages imported from `.docx` files.
I will do further experiments with native outline pages.

**Reply by thumDer** [reply]
I've tried hard to replicate the structure of the exported zip, and import my own content, but without any success. The collection validation fails for some reason, but there is no additional information.

Below you can see my folder structure on the left, and the exported (which imports fine) on the right side. What could be the issue here?

![kép](https://user-images.githubusercontent.com/54556708/228488256-cb78c8cc-5cb7-469d-a823-aec4c951966e.png)

**Reply by thumDer** [reply]
It throws the following error: https://github.com/outline/outline/blob/f86ae64a69fdc38c97dd6b8b784685faf6b837e9/server/queues/tasks/ImportTask.ts#L112-L115

So there is no collections returned from `ImportMarkdownZipTask` task.

The same happens for a very simple folder structure, without attachments:
![kép](https://user-images.githubusercontent.com/54556708/228493475-9f6406b1-7af7-4b51-99d3-8c0d7e4fa69a.png)

Unfortunately I cannot debug any further, because we are on the cloud version, but this functionality would really speed up our migration process.

**Reply by thumDer** [reply]
Okay, I've managed to track down the problem. It is not related to the folder structure at all, it is a matter of how the zip file was created. I've opened a related issue: https://github.com/outline/outline/issues/5117

**Reply by tommoor** [reply]
It could be that the native windows utility is adding an extra internal folder to the zip.

**Reply by thumDer** [reply]
I've checked that extra carefully, and the folder structure inside both zip files is exactly the same, so it will be something else. Please find the two version of the same collection structure attached:
[outline-import-win.zip](https://github.com/outline/outline/files/11100459/outline-import-win.zip)
[outline-import-winrar.zip](https://github.com/outline/outline/files/11100460/outline-import-winrar.zip)

**Reply by tommoor** [reply]
Yep

**Comment by WarlockFish** [comment]
hi, I also have this problem. How did you solved it ?

i have this  structured documents as

```
├── assets
│   ├── 1660187562-0422b5ea49766a0296f89903edc050d3.css
│   ├── 1660187562-186f331953090b90478ad634bed372ea.png
│   ├── 1660187562-2799831f76b5383b3e513e0621814052.png
│   └── 1660187562-ec73477221af77a251658f72a83cca6b.png
├── a-title_Linux-free-manage.md
```

the outline export'zip is that

```
├── 标准输入输出.md
├── APUE
│   ├── 第二章.md
│   └── uploads
│       └── 2700b633-bfdb-469f-b98a-4f9c0257906f
│           └── deb971a6-cde9-491e-90f6-b762977d621f
│               └── 图片.png
├── APUE.md
└── uploads
    └── 2700b633-bfdb-469f-b98a-4f9c0257906f
        └── c1ea23ea-c287-4def-8a9b-e46b7c404f02
            └── UNIX环境高级编程（中文第三版）.pdf
```

i don't konw how to  handle the `2700b633-bfdb-469f-b98a-4f9c0257906f` and `deb971a6-cde9-491e-90f6-b762977d621f` How did you solved it ?

**Reply by tommoor** [reply]
Both of the zip files uploaded in the replies above have the correct structure

**Reply by thumDer** [reply]
Just rename your assets folder to `uploads`, the importer searches for that specific folder name. You will need to batch-update your md links. You will need an additional folder level, like:
```
└── My Collection
    ├── uploads
    │   ├── 1660187562-0422b5ea49766a0296f89903edc050d3.css
    │   ├── 1660187562-186f331953090b90478ad634bed372ea.png
    │   ├── 1660187562-2799831f76b5383b3e513e0621814052.png
    │   └── 1660187562-ec73477221af77a251658f72a83cca6b.png
    └── a-title_Linux-free-manage.md
```

**Reply by WarlockFish** [reply]
thank you，i understand :sweat_smile:
